README.md
# #|: Pipe data to and from IRC
[![](http://meritbadge.herokuapp.com/hashpipe)](https://crates.io/crates/hashpipe)

Hashpipe lets you connect stdin and stdout to IRC with Unix pipes. For example,

```
sleep 5; echo "done" | hashpipe --server irc.freenode.net
```

messages the `#hashpipe` channel on the Freenode network when a long-running command finishes.

By default, stdin is echoed to all channels hashpipe joins, and NOTICEs and PRIVMSGs are printed to stdout. You can even use it as a (poor) IRC client:

```
$ hashpipe -s irc.freenode.net
lm->#hashpipe: hello there hashpipe
hello to you, lm
lm->hashpipe: I can even /msg you
neat!
(I can't msg you back, though)
```

Hashpipe is also handy for snooping on IRC channels:

```
if hashpipe --server my.irc.server --channels "#commandcentre" | grep -m 1 "oh geez, it's the law"; then
  shred ~/Documents/secret/**/*
fi
```

You can even do semi-interactive things with hashpipe:

```
echo "Say 'stop' in the next 60 seconds to cancel a reboot" | timeout 60 hashpipe -s my.irc.server | grep -m 1 "stop"
ret=$?
if [[ $ret -eq 1 ]]; then
  reboot;
fi
```

Of course, the best application is sending channel spam^W^Wimportant messages to channels:

```
$ echo "srs bznz" | figlet -f slant | hashpipe -s irc.for.professionals -c "#important"
```

```
<hashpipe>                       __
<hashpipe>    _______________   / /_  ____  ____  ____
<hashpipe>   / ___/ ___/ ___/  / __ \/_  / / __ \/_  /
<hashpipe>  (__  ) /  (__  )  / /_/ / / /_/ / / / / /_
<hashpipe> /____/_/  /____/  /_.___/ /___/_/ /_/ /___/
```

## Raw mode
Hashpipe can also parse stdin as raw IRC commands and/or output everything the server sends. This is primarily useful for sending direct PRIVMSGs or automating various oper tasks.

It understands how to parse the commands listed [here](http://aatxe.github.io/irc/irc/client/data/command/enum.Command.html).
If a command fails to parse, it will warn you, but continue processing stdin.

For instance,
```
$ hashpipe --server my.irc.server -io
:my.irc.server 1 hashpipe :Welcome to My IRC Server hashpipe!hashpipe@213.456.8.92
<snip>
:hashpipe!hashpipe@213.456.8.92 MODE hashpipe +x
JOIN #hashpipe
:hashpipe!hashpipe@213.456.8.92 JOIN #hashpipe
:my.irc.server 353 hashpipe @ #hashpipe :hashpipe @`lm` 
:my.irc.server 366 hashpipe #hashpipe :End of /NAMES list.
PRIVMSG #hashpipe :hello
PRIVMSG lm :hi there
NICK bob
:hashpipe!hashpipe@213.456.8.92 NICK :bob
```

## Usage
```
USAGE:
    hashpipe [FLAGS] [OPTIONS] --server <server>

FLAGS:
    -h, --help       Prints help information
    -d, --no-ssl     Disable SSL encryption
    -q               Only print errors (overrides -v; overridden by raw output)
    -i, --raw-in     Interpret STDIN as raw IRC commands
    -o, --raw-out    Echo everything from the IRC server directly
    -v               Verbosity (1 for info, 2 for debug)
    -V, --version    Prints version information

OPTIONS:
    -c, --channels <channels>    Channel(s) to speak in (defalt: #hashpipe, or nothing if using raw input)
    -n, --nick <nick>            Nickname to use (default: hashpipe)
    -p, --port <port>            Port to use (default: 6667, or 6697 with SSL)
    -s, --server <server>        IRC server to connect to
```



CHANGELOG.md
## 0.1.0: Initial Release
Hashpipe was rudely released upon an unsuspecting world.

### 0.1.1: Connection bugfix
When piping input in in raw-input mode, hashpipe was too eager to send its stdin to the server. Now it patiently waits for the server to listen before speaking.

### 0.1.2: Bump `clap` version
The current version of `clap` would not build on new versions of `rustc`.

src/main.rs
#[macro_use]
extern crate chan;
extern crate chan_signal;
use chan_signal::Signal;
use std::thread::spawn;

#[macro_use]
extern crate clap;

#[macro_use]
extern crate log;
extern crate env_logger;
use log::LogLevelFilter;
use env_logger::LogBuilder;

use std::convert::From;

extern crate irc;
use irc::client::prelude::*;
use std::default::Default;
use std::str::FromStr;

extern crate isatty;
use isatty::stdout_isatty;

use std::io::prelude::*;
use std::io::BufReader;
use std::io::Error as IoError;
use std::io::stdout;


/// Actions that threads can take
#[derive(Debug)]
enum Action {
    Connect,
    Quit,
    Join,
    Continue,
    JoinFail(String),
    IoError(IoError),
    ParseError(&'static str),
}

impl From<IoError> for Action {
    fn from(err: IoError) -> Action {
        Action::IoError(err)
    }
}

impl From<&'static str> for Action {
    fn from(err: &'static str) -> Action {
        Action::ParseError(err)
    }
}


fn main() {
    // Catch signals we expect to exit cleanly from
    // I don't think rust actually sends SIGPIPE, but we'll catch it anyway
    let signal = chan_signal::notify(&[Signal::INT, Signal::TERM, Signal::PIPE]);


    // Parse args
    let matches = clap_app!
        (hashpipe =>
         (version: crate_version!())
         (author: crate_authors!())
         (about: "#|: Pipes data to and from an IRC connection")
         (@arg server: -s --server +required +takes_value "IRC server to connect to")
         (@arg port: -p --port +takes_value "Port to use (default: 6667, or 6697 with SSL)")
         (@arg no_ssl: -d long("--no-ssl") "Disable SSL encryption")
         (@arg nick: -n --nick +takes_value "Nickname to use (default: hashpipe)")
         (@arg channels: -c --channels +takes_value "Channel(s) to speak in \
          (defalt: #hashpipe, or nothing if using raw input)")
         // NOTE: long() is a required workaround for parsing long options with
         // hyphens; see https://github.com/kbknapp/clap-rs/issues/321
         (@arg raw_out: -o long("--raw-out") "Echo everything from the IRC server directly")
         (@arg raw_in: -i long("--raw-in") "Interpret STDIN as raw IRC commands")
         (@arg v: -v +multiple "Verbosity (1 for info, 2 for debug)")
         (@arg quiet: -q "Only print errors (overrides -v; overridden by raw output)")
        )
        .get_matches();

    let raw_out = matches.is_present("raw_out");
    let raw_in = matches.is_present("raw_in");
    let quiet = matches.is_present("quiet");

    let nick = matches.value_of("nick").unwrap_or("hashpipe").to_string();
    let server = matches.value_of("server").unwrap().to_string();
    let ssl = !matches.is_present("no_ssl");
    let port = matches.value_of("port").and_then(|p| p.parse().ok());

    // Filter empty strings from channel names
    let sanitized_chans: Vec<String> = matches.value_of("channels")
        .map(|chans| {
            chans.split(",")
                .filter(|x| *x != "")
                .map(|x| x.to_string())
                .collect()
        })
        .unwrap_or(vec![]);

    let channels: Vec<String> = if sanitized_chans.len() != 0 || raw_in {
        sanitized_chans
    } else {
        vec!["#hashpipe".to_string()]
    };


    // Set up logger
    let mut builder = LogBuilder::new();
    let level = match (matches.occurrences_of("v"), quiet) {
        (_, true) => LogLevelFilter::Error,
        (0, _) => LogLevelFilter::Warn,
        (1, _) => LogLevelFilter::Info,
        (2, _) | _ => LogLevelFilter::Debug,
    };
    builder.filter(None, level);
    builder.init().unwrap();


    // Set up IRC server
    let cfg = Config {
        nickname: Some(nick),
        server: Some(server),
        port: port,
        use_ssl: Some(ssl),
        channels: Some(channels.clone()),
        ..Default::default()
    };

    let server = match IrcServer::from_config(cfg) {
        Ok(val) => val,
        Err(e) => {
            error!("{}", e);
            return;
        }
    };

    // Connect to IRC on its own thread
    let irc_server = server.clone();
    let (sirc, rirc) = chan::sync(0);

    debug!("Spawning IRC client...");
    spawn(move || run_irc(irc_server, raw_out, quiet, sirc));

    // Wait until we've joined all the channels we need to
    let mut join_count = 0;
    let mut max_joins = channels.len();
    let mut connected = false;
    while !connected || join_count < max_joins {
        chan_select! {
            signal.recv() -> signal => {
                debug!("Received signal {:?}; quitting", signal);
                server.send_quit("#|").unwrap();
                return;
            },
            rirc.recv() -> action => match action {
                Some(Action::Connect) => {
                    connected = true;
                    debug!("Connected!");
                },
                Some(Action::Join) => join_count += 1,
                Some(Action::Quit) => {
                    debug!("QUIT received while attempting to join channels");
                    server.send_quit("#|").unwrap();
                    return;
                },
                Some(Action::JoinFail(err)) => {
                    max_joins -= 1;
                    warn!("{}", err);
                },
                Some(Action::IoError(err)) => {
                    error!("{}", err);
                    server.send_quit("#|").unwrap();
                    return;
                },
                _ => ()
            },
        }
    }

    info!("Joined {} channels", join_count);

    // Open stdin and write it to the desired channels
    let io_server = server.clone();
    let (sio, rio) = chan::sync(0);

    debug!("Spawning stdin reader...");
    spawn(move || run_io(io_server, channels, raw_in, sio));

    loop {
        chan_select! {
            signal.recv() -> signal => {
                debug!("Received signal {:?}; quitting", signal);
                break;
            },
            rio.recv() -> action => match action {
                Some(Action::Continue) => (), // Stop stdin from blocking everything
                Some(Action::IoError(err)) => {
                    error!("{}", err);
                    break;
                },
                Some(Action::ParseError(err)) => {
                    // TODO should this quit?
                    warn!("{}", err);
                },
                _ => if stdout_isatty() {
                    break;
                },
            },
            rirc.recv() -> action => match action {
                Some(Action::Quit) => {
                    debug!("Quit received");
                    break;
                },
                Some(Action::IoError(err)) => {
                    error!("{}", err);
                    break;
                },
                _ => (),
            },
        }
    }

    info!("Quitting!");
    server.send_quit("#|").unwrap();
}

/// Manage IRC connection; read and print messages; signal on JOIN or QUIT
fn run_irc(server: IrcServer, raw: bool, quiet: bool, sjoin: chan::Sender<Action>) {
    server.identify().unwrap_or_else(|err| sjoin.send(From::from(err)));

    for message in server.iter() {
        match message {
            Ok(msg) => {
                if raw {
                    print!("{}", msg);
                    stdout().flush().unwrap_or_else(|err| sjoin.send(From::from(err)));
                }
                match msg.command {
                    Command::JOIN(ref _channel, ref _a, ref _b) => sjoin.send(Action::Join),
                    Command::PRIVMSG(ref target, ref what_was_said) => {
                        if !raw && !quiet {
                            println!("{}->{}: {}",
                                     msg.source_nickname().unwrap_or("* "),
                                     target,
                                     what_was_said);
                            stdout().flush().unwrap_or_else(|err| sjoin.send(From::from(err)));
                        }
                    }
                    Command::NOTICE(ref target, ref what_was_said) => {
                        if !raw && !quiet {
                            println!("{}->{}: {}",
                                     msg.source_nickname().unwrap_or("* "),
                                     target,
                                     what_was_said);
                            stdout().flush().unwrap_or_else(|err| sjoin.send(From::from(err)));
                        }
                    }
                    Command::QUIT(ref _quitmessage) => sjoin.send(Action::Quit),
                    Command::Response(ref response, ref command, ref err) => {
                        // Handle un-joinable channels
                        let the_problem = command.get(1).map_or("", |s| &*s);
                        let errmsg = the_problem.to_string() + ": " +
                                     &err.clone().unwrap_or("".to_string());

                        match *response {
                            Response::ERR_CHANNELISFULL => sjoin.send(Action::JoinFail(errmsg)),
                            Response::ERR_INVITEONLYCHAN => sjoin.send(Action::JoinFail(errmsg)),
                            Response::ERR_BANNEDFROMCHAN => sjoin.send(Action::JoinFail(errmsg)),
                            Response::ERR_BADCHANNELKEY => sjoin.send(Action::JoinFail(errmsg)),
                            Response::ERR_NOSUCHCHANNEL => sjoin.send(Action::JoinFail(errmsg)),

                            Response::RPL_ENDOFMOTD => sjoin.send(Action::Connect),
                            Response::ERR_NOMOTD => sjoin.send(Action::Connect),
                            _ => (),
                        }
                    }
                    _ => (),
                }
            }
            Err(err) => sjoin.send(From::from(err)),
        }
    }

    sjoin.send(Action::Quit)
}

/// Read stdin and write each line to channels/the server
fn run_io(server: IrcServer, channels: Vec<String>, raw: bool, sdone: chan::Sender<Action>) {
    let stdin = BufReader::new(std::io::stdin());
    for line in stdin.lines() {
        match line {
            Ok(ln) => {
                if raw {
                    let raw_line = ln + "\r\n"; // IRC line terminator
                    match Message::from_str(&raw_line) {
                        Ok(msg) => {
                            server.send(msg).unwrap_or_else(|err| sdone.send(From::from(err)))
                        }
                        Err(err) => sdone.send(From::from(err)),
                    }
                } else {
                    for channel in &channels {
                        server.send_privmsg(&channel, &ln)
                            .unwrap_or_else(|err| sdone.send(From::from(err)));
                    }
                }
                sdone.send(Action::Continue);
            }
            Err(err) => sdone.send(From::from(err)),
        }
    }
    // When this function ends, it drops sdone, signaling main
}
